File: client/src/main/java/collection/ChangesApplier.java
package collection;

import model.data.Model;

import java.util.Set;

public class ChangesApplier<T extends Model> {
  private final Set<T> target;

  public ChangesApplier(Set<T> target) {
    this.target = target;
  }

  public void apply(CollectionChangelist<T> changelist) {
    for (CollectionChange<T> change : changelist.getChangelist().values()) {
      applyChange(change);
    }
  }

  private void applyChange(CollectionChange<T> change) {
    target.removeAll(change.getRemoved());
    target.addAll(change.getAdded());
  }
}

File: client-ui/domain/src/commonMain/kotlin/ru.moleus.kollector/domain/collection/ChangesApplier.kt
package ru.moleus.kollector.domain.collection

import collection.CollectionChange
import collection.CollectionChangelist
import model.data.Model

class ChangesApplier<T : Model?>(private val target: MutableSet<T>) {
    fun apply(changelist: CollectionChangelist<T>) {
        for (change in changelist.changelist.values) {
            applyChange(change)
        }
    }

    private fun applyChange(change: CollectionChange<T>) {
        target.removeAll(change.removed)
        target.addAll(change.added)
    }
}